Do hostNet Pod-ports -> hostPorts in Pod defaults

thockin opened the pull request:
Rather than doing it in PodSpec defaulting, which triggers in Deployments and DaemonSets, do it only when a Pod is actually in play.


/kind bug
/kind cleanup
/kind api-change

partial for #117689

```release-note
Pods which set `hostNetwork: true` and declare ports get the `hostPort` field set automatically.  Previously this would happen in the PodTemplate of a Deployment, DaemonSet or other workload API.  Now `hostPort` will only be set when an actual Pod is being created.  If this presents a problem, setting the feature gate "DefaultHostNetworkHostPortsInPodTemplates" to true will revert this behavior.  Please file a kubernetes bug if you need to do this.
```

EDIT by @dims:  Fixed typo - feature name is `DefaultHostNetworkHostPortsInPodTemplates` (not `DefaultHostNetworkHostPortsInWorkloads`)

thockin commented:
/retest

thockin commented:
I thought for sure this would exploide somewhere subtle.  It didn't.  I added tests and a relnote.  This is the part where @liggitt sees some subtle reason why I am a bonehead and can't do this.

liggitt commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -211,10 +215,6 @@ func SetDefaults_PodSpec(obj *v1.PodSpec) {
 	if obj.RestartPolicy == "" {
 		obj.RestartPolicy = v1.RestartPolicyAlways
 	}
-	if obj.HostNetwork {
-		defaultHostNetworkPorts(&obj.Containers)
-		defaultHostNetworkPorts(&obj.InitContainers)
-	}
```
this was an instance of the problems with "derived defaults" described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#static-defaults:

> In some cases, these values may be conditional on or deterministically derived from other fields (e.g. "if otherField is X then this field defaults to 0" or "this field defaults to the value of otherField"). Note that such derived defaults present a hazard in the face of updates - if the "other" field changes, the derived field may have to change, too. The static defaulting logic is unaware of updates and has no concept of "previous value", which means this inter-field relationship becomes the user's problem - they must update both the field they care about and the "other" field.

so changing it (by moving the defaulting to the pod, which is immutable on update) seems useful if we can swing it

liggitt replied:
The immediately mechanical issue I see is that it looks like validation *requires* hostPort == containerPort when hostNetwork == true, even at the deployment, etc, level:
   * ValidatePodTemplateSpec → ValidatePodSpec → ValidatePodSecurityContext → validateHostNetwork: https://github.com/kubernetes/kubernetes/blob/f871d5fabedf346c5db694b726be6dddfd48fa2f/pkg/apis/core/validation/validation.go#L3484-L3492
   * How is this validation passing on deployments that set `hostNetwork: true` but rely on defaulting to populate hostPort in all the container ports? Do we not have tests for that?

More esoteric implications of this:
1. a client that sends Update requests for deployments, etc, will no longer get hostPort fields defaulted inside containers. That means an update to an object persisted in a previous release you would expect to be a no-op will now trigger a generation change and rollout, even though the resulting pods would be identical.
2. pod-creating controllers (replicaset, job, cronjob, statefulset, daemonset) have to tolerate defaulting differences between the pod-spawning object and the pods without triggering rollouts to try to make the pods *exactly* match the workload spec. I *think* that already has to be generally true since we already have pod-specific defaulting, but I've been surprised before, which led to https://github.com/kubernetes/kubernetes/pull/78914, which I don't think we've ever shored up this aspect to verify controllers are well-behaved.

thockin replied:
> it looks like validation requires hostPort == containerPort when hostNetwork == true

And it seems there are no tests for this case.  I can fix that.

> a client that sends Update requests for deployments, etc, will no longer get hostPort fields defaulted inside containers. That means an update to an object persisted in a previous release you would expect to be a no-op will now trigger a generation change and rollout, even though the resulting pods would be identical.

Not sure I follow this one.  A Deployment which already exists has these fields set.  We won't unset them.

> have to tolerate defaulting differences between the pod-spawning object and the pods without triggering rollouts to try to make the pods exactly match the workload spec

I'll look into this, too.  I kicked this PR off to see if it was easy enough to squeeze into free time.  It may not be :)

liggitt replied:
>> a client that sends Update requests for deployments, etc, will no longer get hostPort fields defaulted inside containers. That means an update to an object persisted in a previous release you would expect to be a no-op will now trigger a generation change and rollout, even though the resulting pods would be identical.
>
> Not sure I follow this one. A Deployment which already exists has these fields set. We won't unset them.

When the client sends an _update_ of `{kind:Deployment, spec: {template: {hostNetwork: true, ...}}}`, if we no longer default `hostPort` inside all the container ports, the update would not have those fields populated in the new object on an update. `DeepEqual(newDeployment.spec.template, oldDeployment.spec.template)` would show a diff, generation would get bumped, a new replicaset/etc would roll out

thockin replied:
> the update would not have those fields populated in the new object on an update

 If they did a full replace (PUT) then yes, you are right.  But I am not sure on PATCH - we're into the area of patch/apply where I get fuzzy.

If user sent an update which changed (e.g.) Deployment labels (which should not trigger a redeploy), wouldn't the patch/apply logic carry the hostPort values forward?

liggitt replied:
yes, patch should leave the defaulted fields alone

thockin replied:
Pushed with validation tests.  Will need to look at the rest later

thockin replied:
> I think that already has to be generally true since we already have pod-specific defaulting, but I've been surprised before, which led to  #78914, which I don't think we've ever shored up this aspect to verify controllers are well-behaved.

Yeah That test doesn't handle conditional defaults.  I'll have to comprehend the tangle behind that thing.

thockin replied:
oof.  I think I get it, but it's tricky - it detects a slice with a value as "already defaulted".  I'm pushing an update now - PTAL.  Still need to verify manually that controllers don't hotloop on this.

thockin replied:
I verified by inspection DaemonSet, ReplicaSet, StatefulSet, and Job.  I also ran a Deployment under this config and verified that Deployment's hostPort is not set, the pod's hostPort *is* set, and pods were not being recreated.

liggitt commented:
/hold

aojea commented:
> sees some subtle reason why I am a bonehead and can't do this.

I was 🍿 👀   ... and he never let you down 😅

thockin committed:
Do hostNet Pod-ports -> hostPorts in Pod defaults

Rather than doing it in PodSpec defaulting, which triggers in
Deployments and DaemonSets, do it only when a Pod is actually in play.

aojea commented:
I'm scare of the implications of a change like this, mainly because it always take long time to notice until someone with the edge case reports it ... should we add a feature flag to be able to roll back safely?

thockin commented:
ugh, maybe?  It makes all the tests a lot less pleasant.  I hate you

On Tue, May 2, 2023 at 12:27 AM Antonio Ojea ***@***.***>
wrote:

> I'm scare of the implications of a change like this, mainly because it
> always take long time to notice until someone with the edge case reports it
> ... should we add a feature flag to be able to roll back safely?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/117696#issuecomment-1531008154>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVFBIDVZFKQD7POTR5TXECZMLANCNFSM6AAAAAAXRGEYMY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

thockin commented:
Added a gate, PTAL

thockin commented:
Any more concerns here?

liggitt commented on pkg/features/kube_features.go:
```diff
@@ -986,6 +994,8 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	KubeletTracing: {Default: true, PreRelease: featuregate.Beta},
 
+	LatePodHostnetHostPort: {Default: true, PreRelease: featuregate.Beta},
```
which do we want (impacts the polarity of the naming):
* start enabled and Beta, then graduate to GA, then go away
* start disabled and Deprecated, exist for "a while" until we get evidence it's safe to drop

if we stick with this name, does `Hostnet` read weirdly, rather than `HostNet`?

thockin replied:
Do we have a common pattern of gates meaning "EnableDeprecatedSomething" and trending from true->false as opposed to opposite?  I can't recall seeing one, but I don't object to it if we actually do it.

thockin replied:
Do you have a pref?  I could phrase it either way.  "Enable late defaulting" vs "Disable early defaulting"

liggitt replied:
I'd lean toward something like `DefaultHostPortsInHostNetworkWorkloadObjects`, marked as deprecated, and defaulting to false. Relative terms like early/late seem harder to understand than explicitly saying what gets defaulted

liggitt replied:
(but I only care about 2 out of 10 on the name / stage)

thockin replied:
flipped polarity in next push

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4396,12 +4399,25 @@ func validateSysctls(sysctls []core.Sysctl, fldPath *field.Path) field.ErrorList
-func ValidatePodSecurityContext(securityContext *core.PodSecurityContext, spec *core.PodSpec, specPath, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+// validatePodSecurityContext verifies the SecurityContext of a Pod, but only
+// when it is a Pod and not an embedded PodSpec.
+func validatePodSecurityContext(securityContext *core.PodSecurityContext, spec *core.PodSpec, specPath, scPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
This name is pretty misleading, since this doesn't actually verify the whole security context, only the hostnetwork port bits, and relies on validatePodSpecSecurityContext being called separately. With a single caller, I'd probably just inline this into `validatePodMetadataAndSpec`:

```
	if securityContext != nil {
		allErrs = append(allErrs, validateHostNetwork(securityContext.HostNetwork, spec.Containers, specPath.Child("containers"))...)
	}
```

edit: after reviewing above, we still need to do *some* validation on hostnetwork ports in workload specs, so we may not need the separate function _or_ the inline if we plumb host port validation options down to a shared validation method. See comment above.

thockin replied:
I was trying to avoid retooling all of this , but ACK that it is lacking.  Will take a look at another commit

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3793,7 +3796,7 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 	allErrs = append(allErrs, validateDNSPolicy(&spec.DNSPolicy, fldPath.Child("dnsPolicy"))...)
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabels(spec.NodeSelector, fldPath.Child("nodeSelector"))...)
-	allErrs = append(allErrs, ValidatePodSecurityContext(spec.SecurityContext, spec, fldPath, fldPath.Child("securityContext"), opts)...)
+	allErrs = append(allErrs, validatePodSpecSecurityContext(spec.SecurityContext, spec, fldPath, fldPath.Child("securityContext"), opts)...)
```
going into slow-motion skew visualization mode...

1. This relaxes validation in 1.28 by not checking hostPort values at all when validating workloads, which goes too far.
    * We want to tolerate a hostPort of 0 to allow late-defaulting, but we _don't_ want to tolerate `HostPort != 0 && HostPort != ContainerPort`.
    * I think we still need to validate host ports in all cases, but tolerate zero values when validating workloads and forbid them when validating pods. That can be plumbed as a `RequireHostPortMatchContainerPort bool` in `PodValidationOptions`, and only set to true in ValidatePodCreate/ValidatePodUpdate.

2. Normally we have to roll out validation changes over two releases, _but_ this is tolerating a zero value a 1.27 API server will default, so I _think_ we're ok
    * If a 1.28 server persists a workload with hostPort=0, a 1.27 server will default `hostPort` to match `containerPort` when reading from etcd, passing validation and avoiding a generation bump.
    * If a client reads a workload with hostPort=0 from a 1.28 server, and writes to a 1.27 server, the 1.27 server will default `hostPort` to match `containerPort` when reading from etcd, passing validation and avoiding a generation bump.

thockin replied:
New commit pushed with validation enhanced.

liggitt commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -211,9 +218,11 @@ func SetDefaults_PodSpec(obj *v1.PodSpec) {
-	if obj.HostNetwork {
-		defaultHostNetworkPorts(&obj.Containers)
-		defaultHostNetworkPorts(&obj.InitContainers)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.LatePodHostnetHostPort) {
```
if we go with an off-by-default `Deprecated` feature gate name, this would switch polarity, e.g. `if utilfeature.DefaultFeatureGate.Enabled(features.DefaultHostPortsInHostNetworkWorkloadObjects)`

thockin committed:
Retool validation for pod HostNetwork ports

This will ensure that HostPort == ContainerPort for pods and that
HostPort == 0 || HostPort == ContainerPort for embedded PodSpecs.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3481,15 +3481,38 @@ func validatePodDNSConfig(dnsConfig *core.PodDNSConfig, dnsPolicy *core.DNSPolic
 				idxPath := portsPath.Index(i)
-				if port.HostPort != port.ContainerPort {
-					allErrors = append(allErrors, field.Invalid(idxPath.Child("containerPort"), port.ContainerPort, "must match `hostPort` when `hostNetwork` is true"))
+				if hp, cp := port.HostPort, port.ContainerPort; (hp != 0 || opts.ResourceIsPod) && hp != cp {
```
`hp != cp` looks like a duplicate of the `if port.HostPort != port.ContainerPort` check below?

either simplify the outer if or drop the inner if?

thockin replied:
dangit, multiple iterations and I still left debris.  At least I removed the `fmt.Printf("TIM: 1")` before you saw it.

liggitt commented on pkg/features/kube_features.go:
```diff
@@ -212,6 +212,15 @@ const (
 	CronJobTimeZone featuregate.Feature = "CronJobTimeZone"
 
+	// owner: @thockin
+	// beta: v1.28
```
```suggestion
	// deprecated: v1.28
```

thockin replied:
ACK

liggitt reviewed (commented):
/lgtm
/approve

/hold
logic looks correct, two small comments

liggitt reviewed (commented):
/lgtm
/approve

/hold
logic looks correct, two small comments

thockin commented:
re-pushed

aojea commented on pkg/features/kube_features.go:
```diff
@@ -212,6 +212,15 @@ const (
 	CronJobTimeZone featuregate.Feature = "CronJobTimeZone"
 
+	// owner: @thockin
+	// deprecated: v1.28
```
oh, this is much better than current method we were doing

aojea replied:
I see https://github.com/kubernetes/kubernetes/pull/117696#discussion_r1189210781

thockin replied:
Yeah, I didn't realize this was an option either :)

akhilerm commented:
Is the issue https://github.com/kubernetes/kubernetes/issues/126879 caused by this change?
